worktree.go

```diff
@@ -910,8 +1037,43 @@ func (w *Worktree) Submodules() (Submodules, error) {
 		return nil, err
 	}
 
+	// Pre-fetch origin remote for relative URL resolution to avoid repeated lookups
+	var originURL string
+	for _, s := range m.Submodules {
+		if s.URL != "" && isRelativeURL(s.URL) {
+			originURL, err = w.getOriginRemoteURL()
+			if err != nil {
+				return nil, err
+			}
+			break
+		}
+	}
+
 	for _, s := range m.Submodules {
-		l = append(l, w.newSubmodule(s, c.Submodules[s.Name]))
+		if s.URL == "" || !isRelativeURL(s.URL) {
+			l = append(l, w.newSubmodule(s, c.Submodules[s.Name]))
+			continue
+		}
+
+		// Create a copy to avoid modifying the original submodule data
+		copiedSubmodule := *s
+
+		var copiedConfig *config.Submodule
+		if c.Submodules[s.Name] != nil {
+			tempConfig := *c.Submodules[s.Name]
+			copiedConfig = &tempConfig
+		}
+
+		resolvedURL, err := w.resolveRelativeSubmoduleURL(s.URL, originURL)
+		if err != nil {
+			return nil, fmt.Errorf("failed to resolve relative submodule URL %q: %w", s.URL, err)
+		}
+		copiedSubmodule.URL = resolvedURL
+		if copiedConfig != nil {
+			copiedConfig.URL = resolvedURL
+		}
+
+		l = append(l, w.newSubmodule(&copiedSubmodule, copiedConfig))
```

Comment:
Making `resolveModuleURL` the source of truth, means that we don't need to have any logic here around whether or not the URL is relative:

```suggestion
		sub := w.newSubmodule(s, c.Submodules[s.Name])
		cfg := sub.Config()

		resolvedURL, err := resolveModuleURL(originURL, cfg.URL)
		if err != nil {
			return nil, fmt.Errorf("failed to resolve submodule URL %q: %w", s.URL, err)
		}

		cfg.URL = resolvedURL
		l = append(l, sub)
```